### PR TITLE
Introduce `initialize()` method in `AbstractRoutingDataSource` and `AbstractRoutingConnectionFactory`

### PR DESCRIPTION
**Pull request by violetbeach**
Situation is that you want to add DataSource to the DataSources of AbstractRoutingDataSource during runtime.

When i put a Data Source, I need to synchronize target Data Sources to resolved Data Sources.

At that time, I call `afterPropertiesSet()` as below.

```java
@Component
public class MultiDatabaseManager {

    private final Map<Object, Object> dataSourceMap = new ConcurrentHashMap<>();

    private AbstractRoutingDataSource multiDataSource;

    public DataSource createMultiDataSource() {
        MultiDataSource multiDataSource = new MultiDataSource();
        multiDataSource.setTargetDataSources(dataSourceMap);
        multiDataSource.setDefaultTargetDataSource(defaultDataSource());
        return multiDataSource;
    }

    public void addDataSource(String hostIp, ...) throws SQLException {
        DataSource dataSource = DataSourceBuilder
                .create()
                .username(username)
                .password(password)
                .url(url)
                .driverClassName(driverClassName)
                .build();

        try (Connection c = dataSource.getConnection()) {
            dataSourceMap.put(hostIp, dataSource);
            multiDataSource.afterPropertiesSet(); // unclear intentions
        }
    }

}
```

However, `afterPropertiesSet()` is unclear.

So I suggest extract `refresh()` method. Then, intent can be exposed.

Please feel free to share your opinions or advice.

thanks you.

### CODE REVIEW

**Review by sbrannen — CHANGES_REQUESTED** [review]
Thanks for the proposal and PR.

Can you please also introduce a similar `refresh()` method in `AbstractRoutingConnectionFactory` along with a corresponding test?

**Comment by violetbeach** [comment]
> Can you please also introduce a similar `refresh()` method in `AbstractRoutingConnectionFactory` along with a corresponding test?

Sure. I've done.

Please let me know if you need anything else. Thank you for your review!

File: spring-jdbc/src/main/java/org/springframework/jdbc/datasource/lookup/AbstractRoutingDataSource.java [hunk 1]
```diff
@@ -116,6 +116,15 @@ public void setDataSourceLookup(@Nullable DataSourceLookup dataSourceLookup) {
...
 	public void afterPropertiesSet() {
+		refresh();
+	}
+
+	/**
+	 * Synchronizes targetDataSources to resolvedDataSources
+	 * and defaultTargetDataSource to resolvedDefaultDataSource.
+	 * @throws IllegalArgumentException in case of targetDataSources is null
+	 */
+	public void refresh() {
```
Comment (sbrannen) [inline spring-jdbc/src/main/java/org/springframework/jdbc/datasource/lookup/AbstractRoutingDataSource.java:127]: Please rename to `initialize()`.

File: spring-r2dbc/src/main/java/org/springframework/r2dbc/connection/lookup/AbstractRoutingConnectionFactory.java [hunk 2]
```diff
@@ -127,6 +127,14 @@ public void setConnectionFactoryLookup(ConnectionFactoryLookup connectionFactory
...
 	@Override
 	public void afterPropertiesSet() {
+		refresh();
+	}
+
+	/**
+	 * Synchronizes targetConnectionFactories to resolvedConnectionFactories
+	 * and defaultTargetConnectionFactory to resolvedDefaultConnectionFactory.
+	 */
+	public void refresh() {
```
Comment (sbrannen) [inline spring-r2dbc/src/main/java/org/springframework/r2dbc/connection/lookup/AbstractRoutingConnectionFactory.java:137]: Please rename to `initialize()`.

**Review by sbrannen — CHANGES_REQUESTED** [review]
Thanks for making the changes to `AbstractRoutingConnectionFactory` as well.

After further consideration, the team has decided that we would like these new methods to be named `initialize()` instead of `refresh()`, since the latter implies that there is first class support for dynamic refreshes at runtime -- which is not the case.

**Comment by violetbeach** [comment]
> After further consideration, the team has decided that we would like these new methods to be named `initialize()` instead of `refresh()`, since the latter implies that there is first class support for dynamic refreshes at runtime -- which is not the case.

I get it.

Thank you for considering it and the kind comment!

**Review by sbrannen — APPROVED** [review]

**Comment by sbrannen** [comment]
This has been merged into `main` in 6d2d8a36c2b608e0c127efa20ef9ea8f8238ff84 and revised in d50ec68ad7bbbc195353e1ba59e2a613bc80013a.

Thanks for the contribution, and congratulations on getting your first PR for the Spring Framework merged! 👍

**Pull request closed without merging**